src/cargo/util/toml/targets.rs

```diff
@@ -928,6 +932,54 @@ fn validate_unique_build_scripts(scripts: &[String]) -> CargoResult<()> {
     Ok(())
 }
 
+// Will check if `path` specified for a target is a valid source file
+fn validate_target_path_as_source_file(
+    path: &Path,
+    target_name: &str,
+    target_kind: &str,
+) -> CargoResult<()> {
+    if !path.exists() {
+        anyhow::bail!(
+            "can't find {} `{}` at path `{}`",
+            target_kind,
+            target_name,
+            path.display()
+        );
+    }
```

Comment:
We should have a test for this case as well.  This might also represent a breaking change.